source-dropbox/main.go

```diff
@@ -0,0 +1,300 @@
+package main
+
+import (
+	"cmp"
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"slices"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/estuary/connectors/filesource"
+	"github.com/estuary/flow/go/parser"
+	pf "github.com/estuary/flow/go/protocols/flow"
+)
+
+type config struct {
+	Path        string         `json:"path"`
+	Credentials *Credentials   `json:"credentials"`
+	Parser      *parser.Config `json:"parser"`
+	MatchKeys   string         `json:"matchKeys,omitempty"`
+	Advanced    advancedConfig `json:"advanced"`
+}
+
+type advancedConfig struct {
+	AscendingKeys bool `json:"ascendingKeys,omitempty"`
+}
+
+func (c config) Validate() error {
+	requiredProperties := [][]string{
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	oauthClient := cfg.Credentials.GetOauthClient(ctx)
+	config := dropbox.Config{
+		Client:   oauthClient,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks if the specified path exists in the Dropbox store.
+// It returns an error if the path does not exist or if there was an error
+// while checking the path.
+func (dbStore *dropboxStore) check() error {
+	_, err := dbStore.client.ListFolder(&files.ListFolderArg{Path: dbStore.config.Path, Limit: 1})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (dbStore *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	// Get all files at the specified path from Dropbox.
+	arg := &files.ListFolderArg{Path: query.Prefix, Recursive: query.Recursive}
+	f, err := dbStore.client.ListFolder(arg)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving folder: %w", err)
+	}
+
+	entries := f.Entries
+
+	for f.HasMore {
```

Comment:
A comment here about why we need to do a full listing and sort in-memory would be helpful

Reply:
Good idea. What do you think about adding this comment? I could put it as the first comment within the `List` method.
> The Dropbox API returns pages of entries sorted by their modification date and time. However for the connector to work with the `filesource` package correctly, the listing returned by `List` must contain all entries sorted in lexicographic, monotonic order. To achieve this, we paginate through all entries, convert them to `filesource.ObjectInfo`s, and sort them based on their path.

Reply:
Yeah that seems good. The bit about how the API returns entries sorted differently is great.

Reply:
I'll add that comment. I'll also clarify the comments for the pagination since it's not obvious that the Dropbox API's `ListFolderContinue` fetches the next page of entries.